### PR TITLE
streamingccl: introduce a replication builtin crdb_internal.replication_stream_progress

### LINKED ISSUE #70925: changefeedccl: Implement replication heartbeat mechanism
Implement `crdb_internal.replication_stream_progress(id, ts)` as described in
https://docs.google.com/document/d/1i7x6ce0SGMbJ3KEszeRVjRS0LaPhxulcUbi029nVy7c/edit#

This function is responsible for advancing stream job state on the source cluster; it functions as a way for the destination cluster to advance the PTS record forward, and to let the producer know that replication still happening.

Epic CRDB-6407

### PR DESCRIPTION
**Pull request by gh-casper**
Introduce a new builtin crdb_internal.replication_stream_progress
that can be used on the consumer side to report replication progress
to the source cluster and keeps the stream alive.

crdb_internal.replication_stream_progress(stream_id, frontier_ts):
stream_id is the id for the replication stream.
frontier_ts is the frontier timestamp until which the consumer has
made changes durable.

Fixes: #70925

Release note (sql): introduce a replication builtin
crdb_internal.replication_stream_progress(stream_id, frontier_ts).

### CODE REVIEW

**Review by miretskiy — CHANGES_REQUESTED** [review]
> Reviewed 4 of 12 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73321)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @gh-casper and @stevendanna)

___
*[-- commits, line 12 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73321#-MpqJjCnEXr7FMhw1nqt):*
s/ntroduce/introduce/
___
*[pkg/ccl/streamingccl/replication_manager.go, line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73321#-MpqJnOw7nv6jeKWkIA6:-MpqJnOw7nv6jeKWkIA7:b-8kfn32) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8ad96b0b2ae8194236035dc227465e90f076c8a/pkg/ccl/streamingccl/replication_manager.go#L49)):*
> ```Go
> }
> 
> // UpdateReplicationStreamProgressHook hooks an UpdateReplicationStreamProgress implementation inside streamingccl package.
> ```

could you wrap long-ish comment?
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 166 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73321#-MpqLnE02bNvSQC7-ouf:-MpqLnE02bNvSQC7-oug:b-fu5twa) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8ad96b0b2ae8194236035dc227465e90f076c8a/pkg/ccl/streamingccl/streamproducer/producer_job.go#L166)):*
> ```Go
> 	ts hlc.Timestamp,
> ) (status jobspb.StreamReplicationStatus, err error) {
> 	err = registry.UpdateJobWithTxn(ctx, jobspb.JobID(streamID), txn, false,
> ```

nit: let's lift `false` argument to
```
const useReadLock = false
```
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 175 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73321#-Mpqg0nQ-1UsSsgF6e4L:-Mpqg0nQ-1UsSsgF6e4M:bfbq3iw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8ad96b0b2ae8194236035dc227465e90f076c8a/pkg/ccl/streamingccl/streamproducer/producer_job.go#L175)):*
> ```Go
> 				status.StreamStatus = jobspb.StreamReplicationStatus_STOPPED
> 				// Skip checking PTS record since it might already be released
> 				return nil
> ```

I'm not sure this is okay -- certainly, job has many more states than the above 2 -- not all of them
mean the job is stopped.  

Thinking about this from the consumer side... The consumer needs to know if the job is running -- that part
is communicated by *not* returning an error here -- just basically, performing PTS management.

We probably also need to communicate the fact that the job is cancelled, reverted, or simply doesn't exist.
Presumably, in that case the consumer restart replication from the last known high watermark.
In the case of job not found -- we would get an error from UpdateJobWithTxn.  We should probably change the jobs
code to return a sentinel `JobNotFoundError` instead of `return errors.Errorf("job %d: not found in system.jobs table", j.ID())`...
However, should not not just return  job not found error.  Instead, we should have our own StreamNotFound error which we
return if job is not found *or* if the job is in the `md.status.Terminal()`

Finally, there is this weird case where the job is not running, but we didn't get job not found or terminal... In that case, we really
can't tell.  We should return an error to the caller indicating that it should retry.

All of the above kinda leads me to believe that perhaps returning a stream status like we do above is not very useful.
The operation `update progress` either succeeds or returns an error.  Since we need to return this error over RPC boundarys, perhaps we should encode error code in the returned status (e.g. UNKNOWN_STREAM_STATUS_RETRY, STREAM_INACTIVE, etc)..
___
*[pkg/jobs/jobspb/jobs.proto, line 110 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73321#-MpqMgi49qgz_hn9rr4U:-MpqMgi49qgz_hn9rr4V:b5momme) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8ad96b0b2ae8194236035dc227465e90f076c8a/pkg/jobs/jobspb/jobs.proto#L110)):*
> ```protobuf
> 
> message StreamReplicationStatus {
> ```

nit: drop extra \n?
___
*[pkg/streaming/utils.go, line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73321#-MpqKhtiCVR3it3RrVQU:-MpqKhtiCVR3it3RrVQV:b-ifdqjd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8ad96b0b2ae8194236035dc227465e90f076c8a/pkg/streaming/utils.go#L28)):*
> ```Go
> func (s StreamID) String() string {
> 	return fmt.Sprintf("%d", s)
> }
> ```

not sure why this is needed... Printing steam id with %d works as well as w/ %s.
___
*[pkg/streaming/utils.go, line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73321#-MpqKovL6pqCIPVeyoV9:-MpqKovL6pqCIPVeyoVA:b-jomk4o) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8ad96b0b2ae8194236035dc227465e90f076c8a/pkg/streaming/utils.go#L44)):*
> ```Go
> 	StartReplicationStream(evalCtx *tree.EvalContext, txn *kv.Txn, tenantID uint64) (StreamID, error)
> 
> 	// UpdateReplicationStreamProgress updates the progress of a replication stream.
> ```

It's a nit... but I think I want to fix this here and on top.

Usually, txn is the last argument.
context args are first.
So, evalCtx, streamID, frontier, txn would be my preferred argument order.

**Review by gh-casper — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73321)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @stevendanna)

___
*[-- commits, line 12 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73321#-MpqJjCnEXr7FMhw1nqt):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

s/ntroduce/introduce/
</blockquote></details>

Done.
___
*[pkg/ccl/streamingccl/replication_manager.go, line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73321#-MpqJnOw7nv6jeKWkIA6:-MprmuQFB4AUaEZ560jf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8ad96b0b2ae8194236035dc227465e90f076c8a/pkg/ccl/streamingccl/replication_manager.go#L49)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

could you wrap long-ish comment?
</blockquote></details>

Done.
___
*[pkg/ccl/streamingccl/streamproducer/producer_job.go, line 175 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73321#-Mpqg0nQ-1UsSsgF6e4L:-Mpws0qzEKDLNmzl73AR:b-7hst8e) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8ad96b0b2ae8194236035dc227465e90f076c8a/pkg/ccl/streamingccl/streamproducer/producer_job.go#L175)):*
> introduce
I created 4 status which correspond to an action that consumer needs to take. I put these error codes directly into the returned StreamStatus.
___
*[pkg/streaming/utils.go, line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73321#-MpqKhtiCVR3it3RrVQU:-MprnXTv6HGSAL3EWk45:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8ad96b0b2ae8194236035dc227465e90f076c8a/pkg/streaming/utils.go#L28)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

not sure why this is needed... Printing steam id with %d works as well as w/ %s.
</blockquote></details>

Done.
___
*[pkg/streaming/utils.go, line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/73321#-MpqKovL6pqCIPVeyoV9:-Mpro9qeDWguFiLovpU9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8ad96b0b2ae8194236035dc227465e90f076c8a/pkg/streaming/utils.go#L44)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

It's a nit... but I think I want to fix this here and on top.

Usually, txn is the last argument.
context args are first.
So, evalCtx, streamID, frontier, txn would be my preferred argument order.

</blockquote></details>

Done.

File: pkg/ccl/streamingccl/streamproducer/producer_job.go [hunk 1]
```diff
@@ -187,10 +205,30 @@ func updateReplicationStreamProgress(
 			}
 			return nil
 		})
+
+	if jobs.HasJobNotFoundError(err) || testutils.IsError(err, "not found in system.jobs table") {
```
Comment (miretskiy) [inline pkg/ccl/streamingccl/streamproducer/producer_job.go:209]: do we need to depend on testutils?  Are we getting that error message?
Regardless, I think HasJobNotFoundError should handle the case and we shouldn't depend on the testutil package.

**Review by miretskiy — APPROVED** [review]

**Pull request merged**